File: cpp/math/main_2981.cpp
#include <iostream>
#include <algorithm>
#include <set>
#include <cmath>
using namespace std;

int gcd(int a, int b) {
	const int cmp = a % b;
	if (cmp == 0) {
		return b;
	}
	return gcd(b, cmp);
}

int main() {
	int count;
	cin >> count;
	int* num = new int[count];
	for (int i = 0; i < count; ++i) {
		cin >> num[i];
	}
	sort(num, num + count);
	int ans = num[1] - num[0];
	for (int i = 2; i < count; ++i) {
		ans = gcd(ans, num[i] - num[i - 1]);
	}
	set<int> ansSet;
	ansSet.insert(ans);
	for (int i = 2; i <= sqrt(ans); ++i) {
		if (ans % i == 0) {
			ansSet.insert(i);
			ansSet.insert(ans / i);
		}
	}
	for (auto && value : ansSet) {
		cout << value << ' ';
	}
}

File: kotlin/step/math/main_2981.kt
import kotlin.math.sqrt

fun main() {
    val count = readLine()!!.toInt()
    val num = mutableListOf<Int>()
    for (i in 1..count){
        num.add(readLine()!!.toInt())
    }
    num.sort()
    val gcd = num.zipWithNext { a: Int, b: Int -> b - a }.reduce { gcd, n ->
        var a = gcd
        var b = n
        while (b != 0){
            val r = a % b
            a = b
            b = r
        }
        a
    }
    val ans = mutableSetOf<Int>()
    ans.add(gcd)
    for (i in 2..sqrt(gcd.toDouble()).toInt())
        if (gcd % i == 0){
            ans.add(i)
            ans.add(gcd / i)
        }
    ans.sorted().forEach{print("$it ")}
}